### ISSUE TITLE
The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Runtime

### ISSUE THREAD

**Issue by haf** [dotnet/cli#9337]
## Steps to reproduce on macOS

```
git clone https://github.com/masroore/CurlSharp.git
cd CurlSharp
export FrameworkPathOverride=$(dirname $(which mono))/../lib/mono/4.6.1-api
dotnet restore
dotnet build
... =>
NativeMethods.cs(49,36): error CS0012: The type 'Object' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Runtime, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [/Users/h/dev/logibit/CurlSharp/CurlSharp/CurlSharp.csproj]
```

## Expected  behavior

.Net Core was supposed to be cross platform.

## Actual behavior

It's crashing when trying to compile very simple things.

## Environment data
`dotnet --info` output:

```
.NET Command Line Tools (2.1.105)

Product Information:
 Version:            2.1.105
 Commit SHA-1 hash:  141cc8d976

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.13
 OS Platform: Darwin
 RID:         osx.10.13-x64
 Base Path:   /usr/local/share/dotnet/sdk/2.1.105/

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.7
  Build    : 2d61d0b043915bc948ebf98836fefe9ba942be11

```

```
» mono --version                                                                               h@catamorphism
Mono JIT compiler version 5.12.0.226 (2018-02/9824e260f56 Mon Apr 30 18:45:52 EDT 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)
------------------------------------------------------------
```

### Workaround

Don't use .net core to build (not a workaround, this is the bug itself)

**Comment by dasMulli** [comment]
That's currently blocked by https://github.com/Microsoft/msbuild/issues/1504 - the .NET Core MSBuild doesn't understand the layout of mono's reference assemblies.

You can still build using mono's `msbuild`.

There's an alternative hack at https://github.com/dotnet/sdk/issues/335#issuecomment-371484670 you could try or wait for in implementation based on the design in https://github.com/dotnet/designs/pull/33/files?short_path=4c085c4#diff-4c085c464d992b00504305a5996f804c

**Comment by haf** [comment]
I don't see how

> MONO redirect hack

is a good solution, if it's a hack. Neither do I see why there's even close to the requirement to wait for a whole redesign of how things are run and installed on all types of OS-es, when this is a clear and known issue about where reference assemblies are stored. Solution: look in mono reference assemblies from msbuild and then solve it "the grand way".

**Comment by livarcocc** [comment]
Given that the work to support this scenario is already being tracked on different issues, and we are actively working on publishing reference assemblies for full framework to be buildable cross plat, I am going to close this issue.

**Comment by haf** [comment]
@livarcocc I don't see what you're doing that for, when there are already reference assemblies available cross platform?

**Comment by dasMulli** [comment]
@haf mono's reference assemblies aren't the same as the official netfx reference assemblies available on windows. For example they also contain reference assemblies for mono libraries as well as 3rd party libraries. And no reference assemblies for things that mono doesn't implement.

**Comment by haf** [comment]
@dasMulli I believe Mono's been using the netfx assemblies for approximately two years now. Sure, they are not identical, but you have a gaping wound here and therre quickest way to fix it is to use what's available. After all, it's good enough to run on; and this is not an issue with the finer points of reference assemblies; but the contrary; just compiling hello world cross platform.

So in summary:

 - not the same: true but makes no difference for this bug
 - 3rd party libs: if they are not referenced they make no difference
 - no refs for things mono doesn't implement: this is good, because then I get compiler errors as I should be getting when I use things I can't use

**Comment by dasMulli** [comment]
> "3rd party libs": if they are not referenced they make no difference?

not sure about that.. I think msbuild would prefer them over NuGet references, which is fine as long as you run the build output on mono.

> "no refs for things mono doesn't implement": this is good, because then I get compiler errors as I should be getting when I use things I can't use; sounds a bit like the wanted behaviour, no?

I'm not sure. since there is a drive to build things even though you can't really use them, it would be fine to argue "I just want to build a multi-targeting NuGet package with some extra stuff for .net framework, I don't care if it can run on mono" (personally, I wouldn't do it since it means it isn't tested as the tests wouldn't run on mono then).

> the quickest way to fix it is to use what's available
> …
> "already being worked on": it would seem that work has already completed in Mono's msbuild; take that until you finish your own work then?

What you are essentially asking for now is to push a quick fix / hacky workaround into production. I don't think this should be done for the CLI since there are workarounds you can use if you want to go that way. If the CLI were a tool used by a group of professionals then yeah, but as long as it is the entry point for everyone experimenting with .NET Core, I don't think that should be done. (as opposed to mono/msbuild)

**Comment by haf** [comment]
> I think msbuild would prefer them over NuGet references

My understanding is that only the surface area/signatures is used, the actual implementation is taken from either nuget or the installed framework.

> I'm not sure. since there is a drive to build things even though you can't really use them, it would be fine to argue "I just want to build a multi-targeting NuGet package with some extra stuff for .net framework, I don't care if it can run on mono" (personally, I wouldn't do it since it means it isn't tested as the tests wouldn't run on mono then).

But this is not what I want; I want to link with Mono's libraries because they are going to be used. I don't care that much about building for Windows.

> What you are essentially asking for now is to push a quick fix / hacky workaround into production.

Why do you consider hard-coding a reference library path a quick fix as opposed to e.g. shipping breaking dlls with proprietary IDEs (System.Net.Http)? How is making builds work for everyone ever a hack? Do you have any idea of how much time MSFT is wasting for us outside, with its indecision of what's a compiler (fsc? msbuild? dotnet?) and what's a nuget (System.*? others?) and what's a framework vs a runtime?

So no, I'm absolutely not suggesting a hack. I'm suggesting adding the actual reference lib paths to dotnet. It can be removed once a superior solution has been found.

>  I don't think this should be done for the CLI since there are workarounds you can use if you want to go that way.

No there are no workarounds; that's why I've created this issue. I've tried them all but it keeps crashing on me.

**Comment by dasMulli** [comment]
> But this is not what I want; […] I don't care that much about building for Windows.

Please don't assume that everyone wants the same as you :)

I don't think there is much point in continuing the discussion as it seems to get out of scope.

bottom line:
* You can already build using mono's msbuild
* Hacky workarounds exist to enable a lot of scenarios using the CLI on *nix systems
* Work is already in progress / tracked to enable a proper experience on the CLI side.

I do feel your pain, that's why I opened https://github.com/dotnet/sdk/issues/335 back when I started to play around with the first working CLI-using-msbuild bits and like that MS is going to implement it

**Comment by haf** [comment]
> "I don't care if it can run on mono" ... Please don't assume that everyone wants the same as you :)

Right back at you! ;)

> hacky workaround exist

Better to have a hacky workaround in a single place for everyone to stop wasting people's time.

**In summary: bring in the changes of mono's msbuild to dotnet's msbuild to make cross-platform work possible in the short run without delaying any longer**

**Comment by haf** [comment]
Ping @livarcocc — this is still hurting us. What is the official recommendation to get rid of System.Runtime version warnings and work around the issues?

**Comment by haf** [comment]
E.g. `error FS1108: The type 'HttpStatusCode' is required here and is unavailable. You must add a reference to assembly 'System.Net.Primitives, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.`

**Comment by haf** [comment]
@livarcocc What issues are tracking what you're actively working on?

**Comment by haf** [comment]
I don’t see an update or ETA in that thread

--

**Comment by haf** [comment]
@livarcocc When is this due to be fixed?

**Comment by haf** [comment]
Also, is my issue really https://github.com/Microsoft/visualfsharp/pull/5025#issuecomment-392839605?

Is going for a patch release from 200 -> 201 supposed to cause compiler errors?

Can I behave differently to isolate myself from these things going forward?

I don't want to build with msbuild because I have a cloudbuild build pipeline; using both is rather inconvenient: how would you suggest I write this Dockerfile taking your workarounds into account?

```
FROM microsoft/dotnet:2.0.7-sdk-2.1.105 AS build-env
ARG short_sha
ENV SHORT_SHA=$short_sha \
    LISTEN_PORT=8080
ADD ./Api /app
WORKDIR /app
RUN dotnet restore -r linux-x64
RUN dotnet publish  -c Release -r linux-x64

FROM gcr.io/distroless/dotnet
WORKDIR /app
COPY --from=build-env /app/bin/Release/net471 /app/bin
EXPOSE ${LISTEN_PORT}
CMD ["dotnet", "bin/Api.dll"]
```
